processor/spanmetricsprocessor/processor.go

```diff
@@ -243,12 +243,19 @@ func (p *processorImp) Capabilities() consumer.Capabilities {
 // The original input trace data will be forwarded to the next consumer, unmodified.
 func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) error {
 	defer func() {
+		p.lock.Lock()
 		p.reset()
+		p.lock.Unlock()
 	}()
 
```

Comment:
Say there is thread (a) and (b) both calling `ConsumeTraces()`. 

1. Thread (a) obtains write lock and `aggregateMetrics()` and releases lock. 
2. Thread (b) obtains write lock and `aggregateMetrics()`, releases lock and then also goes onto `buildMetrics()`.
3. Now thread (a) calls `buildMetrics()` with the data from both (a) and (b). Specifically in the delta case this would produce incorrect metrics. 

Do you agree? Potentially buildMetrics and aggregateMetrics should be under the same lock? @MovieStoreGuy any thoughts?

Reply:
Yeah, it also makes sense. The processor is stateful and should not be used concurrently. In practice, it should not be used concurrently either based on the design of OTEL Collector.

I would change the lock to make sure the processor can only be executed serially。

Reply:
Regarding this, and from @MovieStoreGuy 's [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075#discussion_r780894362):

Locks protect data. If there is no data to protect, you don't need a lock.  If something isn't thread safe, then there's data being accessed, by definition.

If the data is "the entire `processorImp`", that's fine, but it should be documented as such, and locked as such.

Also, if you're saying something isn't thread safe, you don't need a lock.  When you declare something as not thread safe, you're explicitly pushing that concern to the caller - if they want to make it thread safe, they can add their own lock.  If you have a lock (and you use it properly!), it **is** thread safe by definition, because that's what locks give you.

Reply:
> Yeah, it also makes sense. The processor is stateful and should not be used concurrently. In practice, it should not be used concurrently either based on the design of OTEL Collector.
> 
> I would change the lock to make sure the processor can only be executed serially.

I think I agree with the first statement here, that the processor should not be used concurrently, but I'm not sure about the second.  There is nothing in the design of the collector that prevents, prohibits, or even discourages concurrent execution.  Any component can spawn new goroutines at any time, and it should be anticipated that receivers will do so regularly.  While a pipeline is conceptually a serial process, there may be many instantiations of the pipeline concurrently executing and any component in a pipeline may trigger delayed or asynchronous execution of the next step in the pipeline.

I would fully expect that this processor's `ConsumeTraces()` method will be invoked concurrently and thus it should ensure that appropriate measures are taken to ensure data consistency.  If it needs to ensure a consistent view of its state without the potential for changes by other concurrent invocations, a lock should probably be held for the duration of an invocation.

Reply:
Thankyou @Aneurysm9 for clearing this up. Fully in favour of holding the lock for the whole invocation then. 

By the way, this PR has moved to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075 (@Tenaria can you close this one now you're back?)

Reply:
@tiedotguy regarding the lock, it is not protecting the `processorImp()` but to make it synchronized. I've updated the comment and hope that's more explicitly now: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075/files#diff-78f9e7446e0e4fb474786b4b2c262486887bf27479f1f8bc598376d6c3c29b72R72.

But as we discussed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075#discussion_r782682242),  we probably should consider running multiple sub-processor so that each sub-processing itself is synchronized(still locked by a mutex) but these sub-processors can be executed concurrently.